### DISCUSSION TITLE
ArC: Instance type problem

### DISCUSSION THREAD

**Discussion by famod** [quarkusio/quarkus#22144] [Q&A] [answered]
Hi @mkouba & @manovotn!

Another one for you 😉 but a really puzzling one:

Can you think of any reason why this (in a `@QuarkusTest`):
```java
    @Inject
    Instance<JpaRepository<? extends BaseEntity<?>, ?>> allRepositores;
```
provides all `quarkus-spring-data-jpa` repos (as expected) on most systems but on one (my local machine), the instance is always empty?

I ruled out any local inconsistencies and it only goes away when changing it to:
```java
    @Inject
    Instance<JpaRepository<?, ?>> allRepositores;
```

PS: `BaseEntity` is our custom entity superclass.

Doesn't make any sense, or does it?

[a comment by an automated account is not shown]

**Comment by manovotn** [comment, marked as answer]
That looks like a spicy issue indeed @famod :)

> I ruled out any local inconsistencies and it only goes away when changing it to:

Do I understand it correctly that those beans are actually existing and can be selected/injected on your local machine if you change the `Instance` declaration?
If so, then that rules out any issues with bean discovery/registration.

Another thing I can think of (probably not the case, but just to be sure) is removed beans - can you verify that on your local machine those beans do not get marked as removable? Alternatively, just mark all of the impls as `@Unremovable` and try with the original ` Instance<JpaRepository<? extends BaseEntity<?>, ?>>` to see what happens.

**Reply by mkouba** [reply]
> Another thing I can think of (probably not the case, but just to be sure) is removed beans - can you verify that on your local machine those beans do not get marked as removable?

This should be easy to check - run the app in the dev mode and open http://localhost:8080/q/dev/io.quarkus.quarkus-arc/beans and http://localhost:8080/q/dev/io.quarkus.quarkus-arc/removed-beans.

**Reply by famod** [reply]
@manovotn 
> Do I understand it correctly that those beans are actually existing and can be selected/injected on your local machine if you change the `Instance` declaration?

Correct, injecting a concrete repo (e.g. `UserRepository`) works but the `Instance` stays empty.

> Alternatively, just mark all of the impls as `@Unremovable` and try with the original ` Instance<JpaRepository<? extends BaseEntity<?>, ?>>` to see what happens.

`@Unremovable` doesn't work on interfaces or does it? Spring data repos are just interfaces for which Quarkus generates beans.

@mkouba
> run the app in the dev mode and open http://localhost:8080/q/dev/io.quarkus.quarkus-arc/beans and http://localhost:8080/q/dev/io.quarkus.quarkus-arc/removed-beans

This doesn't work in this case unfortunatly because it's a `QuarkusTest`. The "real" app doesn't use this kind of injection and is basically using all repos explicitly.


Btw, this test is using a `@TestProfile` but it doesn't look suspicious to me. But I'll check without it.

**Reply by manovotn** [reply]
> @Unremovable doesn't work on interfaces or does it? 

Yea that wouldn't work.
A quick hack to verify that it isn't removed - you could for instance keep the test as-is (i.e. make sure there is no explicit injection point with given impl type) and do `CDI.current().getBeanManager().createInstance().select(UserRepository.class).get()`.
Replace `UserRepository` with whatever is the name of the class that implements that interface.
If the bean is found and selected, then it wasn't removed.

**Reply by famod** [reply]
`CDI.current().getBeanManager().createInstance().select(UserRepository.class).get()` does _not_ fail.

The hierarchy is a bit more complex btw:
- `public interface UserRepository extends FindMandatoryByIdRepository<User, UserId>`
- `public interface FindMandatoryByIdRepository<T, ID extends DomainIdentifier> extends JpaRepository<T, ID>`

but again: it only seems to fail on my machine (Ubuntu, WSL2), neither on a Windows host (where the actual change was implemented) nor on Jenkins (CentOS 7).

**Reply by manovotn** [reply]
Thanks for verifying that.

Another random question - what about compilation of sources? Aren't you using a different compiler compared to what the Win host or Jenkins does? Like the one that comes with eclipse IDE.
(just asking since I once saw a very weird and possibly unrelated error on Weld caused by differences in compilation for classes with generics)

**Reply by famod** [reply]
It's all `mvn`, which makes it even stranger.

**Reply by mkouba** [reply]
And JDK versions?

**Reply by famod** [reply]
All the same (Zulu 17.0.1), enforced by maven-enforcer-plugin.

**Reply by manovotn** [reply]
Hmm, no other idea right now.
I'd try to debug the `Instance` resolution process and try to spot what's going wrong locally.

**Reply by mkouba** [reply]
Yup, just add a breakpoint on the line https://github.com/quarkusio/quarkus/blob/main/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/InstanceImpl.java#L235.

**Reply by famod** [reply]
Not reproducible anymore with 2.7.0. 🤷‍♂️

**Reply by mkouba** [reply]
That's good news, right? ;-)

**Reply by famod** [reply]
Sure!
But we tech guys like to know _why_ something doesn't/didn't work, right? So that's a bit disappointing but I'm not really complaining. 😁

**Reply by mkouba** [reply]
Indeed ;-)

**Comment by mkouba** [comment]
CC @geoand

**Comment by geoand** [comment]
Very weird issue...
